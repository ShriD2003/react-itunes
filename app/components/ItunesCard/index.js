/* eslint-disable complexity */
/**
 * ItunesCard Component
 *
 * Displays a card for an iTunes song with optional detailed view.
 *
 * @component
 * @example
 * const song = {
 *   trackId: 123,
 *   artworkUrl100: 'https://example.com/artwork.jpg',
 *   trackName: 'Song Title',
 *   trackPrice: 1.99,
 *   previewUrl: 'https://example.com/preview.mp3'
 * };
 * return <ItunesCard song={song} trackDetails={true} />;
 */

import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { colors } from '@app/themes';

import { Card, Typography, Button, Progress, Spin } from 'antd';

import { Link } from 'react-router-dom';

import { translate } from '@app/utils';
import { PlayCircleTwoTone, PauseCircleTwoTone } from '@ant-design/icons';

const { Paragraph } = Typography;

const CustomCard = styled(Card)`
  && {
    width: 300px;
    height: 450px;
    margin: 1rem 0;
    padding: 1rem;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    background-color: ${colors.musicGridBg};
    border: 1px solid #ffdbac;
  }
`;

const CardContent = styled.div`
  flex: 1;
`;

const StyledImage = styled.img`
  width: 100%;
  height: auto;
  max-height: 150px;
  object-fit: cover;
  margin-bottom: 1rem;
`;

const StyledLink = styled(Link)`
  display: block;
  margin-bottom: 1rem;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 1rem;
`;

const SpinnerContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 1rem;
`;

/**
 * Card for itunes. This card is used to show trackImage and track details.
 *
 *
 * @return { ReactElement } The card thatll be displayed
 */
function ItunesCard({ song, trackDetails, onActionClick }) {
  const { trackName, artworkUrl100, trackId, previewUrl, artistName } = song;

  const [play, setPlay] = useState(false);
  const [loading, setLoading] = useState(false);
  const [duration, setDuration] = useState(0);
  const [currentTime, setCurrentTime] = useState(0);
  const songElement = useRef(null);

  useEffect(() => {
    const audio = songElement.current;
    if (audio) {
      const setAudioData = () => {
        setDuration(audio.duration);
        setLoading(false);
      };

      const setAudioTime = () => {
        setCurrentTime(audio.currentTime);
      };

      audio.addEventListener('loadedmetadata', setAudioData);
      audio.addEventListener('timeupdate', setAudioTime);
      audio.addEventListener('ended', handleEnded);

      return () => {
        audio.removeEventListener('loadedmetadata', setAudioData);
        audio.removeEventListener('timeupdate', setAudioTime);
        audio.removeEventListener('ended', handleEnded);
      };
    }
  }, []);

  const handleMusic = () => {
    const audio = songElement.current;
    if (play) {
      audio.pause();
    } else {
      if (audio.readyState >= 3) {
        audio.play();
      } else {
        setLoading(true);
        audio.load();
      }
    }
    setPlay(!play);
    onActionClick(songElement);
  };

  const handleEnded = () => {
    setPlay(false);
  };

  const handlePlaybackError = () => {
    console.error('Playback error');
    setPlay(false);
  };

  const truncateText = (text, wordLimit) => {
    const words = text?.split('');
    if (words?.length <= wordLimit) {
      return text;
    }
    return `${words?.slice(0, wordLimit).join('')}...`;
  };
  const truncatedTrackName = truncateText(trackName, 18);
  const truncatedLongDescription = truncateText(song.longDescription, 1);
  const truncatedShortDescription = truncateText(song.shortDescription, 25);

  return (
    <CustomCard data-testid="song-card">
      <CardContent>
        <StyledImage src={artworkUrl100} alt={trackName} />
        <StyledLink to={`/details/${trackId}`}>
          <Typography.Title level={4}>{truncatedTrackName}</Typography.Title>
          <Typography.Text>{artistName}</Typography.Text>
        </StyledLink>
        <Paragraph data-testid="para-test">{!truncatedShortDescription || !truncatedLongDescription || ''}</Paragraph>
      </CardContent>
      <div>
        <ButtonContainer>
          <Button
            data-testid="play-btn"
            onClick={handleMusic}
            disabled={loading}
            type={play ? 'text' : 'ghost'}
            icon={<PlayCircleTwoTone />}
          >
            {translate('play-btn')}
          </Button>
          <Button
            data-testid="stop-btn"
            onClick={handleMusic}
            disabled={!play}
            type={!play ? 'text' : 'ghost'}
            icon={<PauseCircleTwoTone />}
            size="large"
          >
            {translate('stop-btn')}
          </Button>
        </ButtonContainer>
        {loading ? (
          <SpinnerContainer>
            <Spin data-testid="spin" />
          </SpinnerContainer>
        ) : (
          <Progress aria-valuenow="50" role="progressbar" percent={Math.round((currentTime / duration) * 100)} />
        )}
      </div>
      <audio data-testid="audio-element" src={previewUrl} ref={songElement} onError={handlePlaybackError}></audio>
    </CustomCard>
  );
}

ItunesCard.defaultProps = {
  onActionClick: () => {}
};

ItunesCard.propTypes = {
  /** The song data to display in the card */
  song: PropTypes.shape({
    trackId: PropTypes.number,
    artworkUrl100: PropTypes.string,
    trackPrice: PropTypes.number,
    previewUrl: PropTypes.string,
    trackName: PropTypes.string,
    shortDescription: PropTypes.string,
    longDescription: PropTypes.string,
    artistName: PropTypes.string
  }),
  width: PropTypes.number,
  height: PropTypes.number,
  trackDetails: PropTypes.bool,
  onActionClick: PropTypes.func
};

export default ItunesCard;
