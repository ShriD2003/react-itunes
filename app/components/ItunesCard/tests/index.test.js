/**
 *
 * Tests for ItunesCard
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderProvider } from '@utils/testUtils';
import ItunesCard from '../index';

describe('<ItunesCard/> tests', () => {
  let song;
  beforeEach(() => {
    song = {
      trackName: 'Track',
      trackId: 1234,
      shortDescription: 'Song Description',
      artworkUrl100:
        'https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk',
      artistName: 'Artist Name',
      previewUrl: 'https://mockurl.com/'
    };
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ItunesCard song={song} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 SongCard component', () => {
    const { getAllByTestId } = renderProvider(<ItunesCard song={song} />);
    expect(getAllByTestId('song-card').length).toBe(1);
  });
});
