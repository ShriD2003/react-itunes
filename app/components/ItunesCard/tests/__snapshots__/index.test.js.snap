// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItunesCard/> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered css-zsbyvt evxfas45 css-dev-only-do-not-override-1pg9a38"
      data-testid="song-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="css-82a6rk evxfas44"
        >
          <img
            alt="Track"
            class="css-1wylkl evxfas43"
            src="https://www.google.com/url?sa=i&url=https%3A%2F%2Fdribbble.com%2Fshots%2F14395014-Music-Logo&psig=AOvVaw3I_T0J8_ZRKW_g4VF_KmKz&ust=1630933127025000&source=images&cd=vfe&ved=2ahUKEwiT5uO-8efyAhWVHLcAHVSbCvQQjRx6BAgAEAk"
          />
          <a
            class="css-arogqd evxfas42"
            href="/details/1234"
          >
            <h4
              class="ant-typography css-dev-only-do-not-override-1pg9a38"
            >
              Track
            </h4>
            <span
              class="ant-typography css-dev-only-do-not-override-1pg9a38"
            >
              Artist Name
            </span>
          </a>
          <div
            class="ant-typography css-dev-only-do-not-override-1pg9a38"
            data-testid="para-test"
          />
        </div>
        <div>
          <div
            class="css-k7pwy3 evxfas41"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-1pg9a38 ant-btn-ghost"
              data-testid="play-btn"
              type="button"
            >
              <span
                class="ant-btn-icon"
              >
                <span
                  aria-label="play-circle"
                  class="anticon anticon-play-circle"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="play-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                      fill="#1677ff"
                    />
                    <path
                      d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm164.1 378.2L457.7 677.1a8.02 8.02 0 01-12.7-6.5V353a8 8 0 0112.7-6.5l218.4 158.8a7.9 7.9 0 010 12.9z"
                      fill="#e6f4ff"
                    />
                    <path
                      d="M676.1 505.3L457.7 346.5A8 8 0 00445 353v317.6a8.02 8.02 0 0012.7 6.5l218.4-158.9a7.9 7.9 0 000-12.9z"
                      fill="#1677ff"
                    />
                  </svg>
                </span>
              </span>
              <span>
                Play
              </span>
            </button>
            <button
              class="ant-btn css-dev-only-do-not-override-1pg9a38 ant-btn-text ant-btn-lg"
              data-testid="stop-btn"
              disabled=""
              type="button"
            >
              <span
                class="ant-btn-icon"
              >
                <span
                  aria-label="pause-circle"
                  class="anticon anticon-pause-circle"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="pause-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                      fill="#1677ff"
                    />
                    <path
                      d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm-80 524c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V360c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v304zm224 0c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V360c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v304z"
                      fill="#e6f4ff"
                    />
                    <path
                      d="M424 352h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8zm224 0h-48c-4.4 0-8 3.6-8 8v304c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V360c0-4.4-3.6-8-8-8z"
                      fill="#1677ff"
                    />
                  </svg>
                </span>
              </span>
              <span>
                Pause
              </span>
            </button>
          </div>
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="50"
            class="ant-progress ant-progress-status-normal ant-progress-line ant-progress-line-align-end ant-progress-line-position-outer ant-progress-show-info ant-progress-default css-dev-only-do-not-override-1pg9a38"
            role="progressbar"
          >
            <div
              class="ant-progress-outer"
              style="width: 100%;"
            >
              <div
                class="ant-progress-inner"
              >
                <div
                  class="ant-progress-bg ant-progress-bg-outer"
                  style="width: 0%; height: 8px; --progress-percent: 0;"
                />
              </div>
              <span
                class="ant-progress-text ant-progress-text-end ant-progress-text-outer"
                title="0%"
              >
                0%
              </span>
            </div>
          </div>
        </div>
        <audio
          data-testid="audio-element"
          src="https://mockurl.com/"
        />
      </div>
    </div>
  </div>
</body>
`;
